File: app/src/main/java/com/drdisagree/iconify/xposed/ModPack.java
package com.drdisagree.iconify.xposed;

/* Modified from AOSPMods
 * https://github.com/siavash79/AOSPMods/blob/canary/app/src/main/java/sh/siava/AOSPMods/modpacks/XposedModPack.java
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

import android.content.Context;

import de.robv.android.xposed.callbacks.XC_LoadPackage;

public abstract class ModPack {
    protected Context mContext;

    public ModPack(Context context) {
        mContext = context;
    }

    public abstract void updatePrefs(String... Key);

    public abstract boolean listensTo(String packageName);

    public abstract void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) throws Throwable;
}

File: app/src/main/java/com/drdisagree/iconify/xposed/ModPack.kt
package com.drdisagree.iconify.xposed

import android.content.Context
import android.content.pm.PackageManager
import com.drdisagree.iconify.BuildConfig
import io.github.libxposed.api.XposedModuleInterface.PackageReadyParam

abstract class ModPack(private val context: Context) {

    protected val mContext: Context get() = context

    protected val appContext: Context
        get() = try {
            context.createPackageContext(
                BuildConfig.APPLICATION_ID,
                Context.CONTEXT_IGNORE_SECURITY
            )
        } catch (exception: PackageManager.NameNotFoundException) {
            throw RuntimeException(exception)
        }

    abstract fun onPreferenceUpdated(vararg key: String)

    @Throws(Throwable::class)
    abstract fun onPackageLoaded(packageReadyParam: PackageReadyParam)
}